File: programmers/[KAKAO 인턴십] 보석쇼핑.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> solution(vector<string> gems) {
    vector<int> answer;
    set<string> target(gems.begin(), gems.end());
    
    int start = 0;
    map<string, int> basket;
    for (int end = 0; end < gems.size(); end++) {
        basket[gems[end]]++;
        if (basket.size() == target.size()) {
            do {
                if (--basket[gems[start]] == 0) {
                    basket.erase(gems[start]);
                    if (answer.size() == 0 || answer[1] - answer[0] > end - start) {
                        answer.clear();
                        answer.push_back(start + 1);
                        answer.push_back(end + 1);
                    }
                    start++;
                    break;
                }
            } while (++start <= end);
        }
    }
    return answer;
}

File: programmers/[KAKAO 인턴십] 보석쇼핑.kt
class Solution {

    fun solution(gems: Array<String>): IntArray {
        val answer = IntArray(2)
        val types = LinkedHashMap<String, Int>()
        gems.toHashSet().forEach { types[it] = 0 }

        var end = -1
        var start = -1
        var minRange = Int.MAX_VALUE
        while (end < gems.size) {
            while (++end < gems.size) {
                types[gems[end]] = types[gems[end]]!! + 1
                if (hasAllGem(types)) break
            }
            while (++start < gems.size) {
                types[gems[start]] = types[gems[start]]!! - 1
                if (!hasAllGem(types)) break
            }
            if (minRange > end - start) {
                minRange = end - start
                answer[0] = start + 1
                answer[1] = end + 1
            }
        }
        return answer
    }

    private fun hasAllGem(types: HashMap<String, Int>): Boolean {
        for (g in types) {
            if (g.value < 1) return false
        }
        return true
    }
}